include/rmm/device_buffer.hpp

```diff
@@ -0,0 +1,325 @@
+   * @param mr Memory resource to use for the device memory allocation
+   *---------------------------------------------------------------------------**/
+  device_buffer(void const* source_data, std::size_t size,
+                cudaStream_t stream = 0,
+                mr::device_memory_resource* mr = mr::get_default_resource())
+      : _size{size}, _capacity{size}, _stream{stream}, _mr{mr} {
+    if (nullptr == source_data) {
+      throw std::runtime_error{"source_data is null."};
+    }
+    _data = _mr->allocate(_size, stream);
+    auto status =
+        cudaMemcpyAsync(_data, source_data, _size, cudaMemcpyDefault, _stream);
+    if (cudaSuccess != status) {
+      throw std::runtime_error{"Copy failed."};
+    }
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Constructs a new `device_buffer` by deep copying the contents of
+   * another `device_buffer`.
+   *
+   * @note Only copies `other.size()` bytes from `other`, i.e., if `other.size()
+   * != other.capacity()`, then the size and capacity of the newly constructed
+   *`device_buffer` will be equal to `other.size()`.
+   *
+   * @throws std::bad_alloc If creating the new allocation fails
+   * @throws std::runtime_error if copying from `other` fails
+   *
+   * @param other The other `device_buffer` to deep copy
+   *---------------------------------------------------------------------------**/
+  device_buffer(device_buffer const& other)
+      : _size{other._size},
+        _capacity{other._size},
+        _stream{other._stream},
+        _mr{other._mr} {
+    _data = _mr->allocate(_size, _stream);
+    auto status =
+        cudaMemcpyAsync(_data, other._data, _size, cudaMemcpyDefault, _stream);
+
+    if (cudaSuccess != status) {
+      throw std::runtime_error{"Device memory copy failed."};
+    }
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Constructs a new `device_buffer` by moving the contents of another
+   * `device_buffer` into the newly constructed one.
+   *
+   * After the new `device_buffer` is constructed, `other` is modified to be
+   * empty, i.e., `data()` returns `nullptr`, and `size()` is zero.
+   *
+   * @throws Nothing
+   *
+   * @param other The `device_buffer` whose contents will be moved into the
+   * newly constructed one
+   *---------------------------------------------------------------------------**/
+  device_buffer(device_buffer&& other) noexcept
+      : _data{other._data},
+        _size{other._size},
+        _capacity{other._capacity},
+        _stream{other._stream},
+        _mr{other._mr} {
+    other._data = nullptr;
+    other._size = 0;
+    other._capacity = 0;
+    other._stream = 0;
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Copy assignment operator
+   *
+   * TODO: Decide if this should be deleted or not.
```

Comment:
What are the pros and cons?

Reply:
It's mainly a case of trying to protect users from themselves. I think it's rare that you'd need copy assignment for a `device_buffer`. Copy construction is usually the better choice if you need to make a copy. By deleting the copy assignment operator, it just prevents a user from unknowingly doing something that is very expensive.